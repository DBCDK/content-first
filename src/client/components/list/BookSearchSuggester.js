import React from 'react';
import request from 'superagent';
import Autosuggest from 'react-autosuggest';
import BookCover from '../general/BookCover.component';

const addEmphasisToString = (string, pattern) => {
  const index = string.toLowerCase().indexOf(pattern.toLowerCase());
  if (index >= 0) {
    const start = index > 0 ? string.slice(0, index) : '';
    const end = string.slice(index + pattern.length);
    const match = string.slice(index, index + pattern.length);
    return (
      <span>
        {start}
        <u>{match}</u>
        {end}
      </span>
    );
  }
  return string;
};

const renderSuggestion = (suggestion, suggestionString) => {
  return (
    <div className="suggestion-row flex">
      <div className="image">
        <BookCover book={suggestion.book} />
      </div>
      <div>
        <div className="suggestion-row__title">
          {addEmphasisToString(suggestion.book.title, suggestionString)}
        </div>
        <div className="suggestion-row__creator">
          {addEmphasisToString(suggestion.book.creator, suggestionString)}
        </div>
      </div>
    </div>
  );
};

class BookSearchSuggester extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || '',
      suggestions: []
    };
  }

  onChange(event, {newValue}) {
    this.setState({
      value: newValue
    });
  }

  async onSuggestionsFetchRequested({value}) {
    this.currentRequest = value;
    const results = JSON.parse(
      (await request.get(
        '/v1/search?q=' +
          encodeURIComponent(
            value
              .trim()
              .split(/\s+/g)
              .join(' & ') + ':*'
          )
      )).text
    ).data.map(book => ({links: {}, book}));
    if (this.currentRequest === value) {
      this.setState({suggestions: results});
    }
  }

  onSuggestionsClearRequested() {
    this.setState({suggestions: []});
    delete this.currentRequest;
  }

  onSuggestionSelected(e, props) {
    e.preventDefault();
    this.props.onSubmit(props.suggestion);
    this.setState({value: ''});
  }
  render() {
    const {value, suggestions} = this.state;

    // Autosuggest will pass through all these props to the input.
    const inputProps = {
      placeholder: 'Skriv titel eller forfatter',
      className: 'form-control suggestion-list__search',
      value,
      onChange: (e, change) => this.onChange(e, change)
    };

    // Finally, render it!
    return (
      <div className="suggestion-list">
        <Autosuggest
          suggestions={suggestions}
          onSuggestionsFetchRequested={e => this.onSuggestionsFetchRequested(e)}
          onSuggestionsClearRequested={() => this.onSuggestionsClearRequested()}
          onSuggestionSelected={(e, props) =>
            this.onSuggestionSelected(e, props)
          }
          getSuggestionValue={({book}) => book.title}
          renderSuggestion={suggestion => renderSuggestion(suggestion, value)}
          inputProps={inputProps}
        />
      </div>
    );
  }
}

export default BookSearchSuggester;
