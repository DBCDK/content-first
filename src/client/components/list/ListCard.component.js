import React from 'react';
import {connect} from 'react-redux';
import TruncateMarkup from 'react-truncate-markup';
import ProfileImage from '../general/ProfileImage.component';
import Link from '../general/Link.component';
import toColor from '../../utils/toColor';
import Heading from '../base/Heading';
import Icon from '../base/Icon';
import SkeletonText from '../base/Skeleton/Text';
import SkeletonUser from '../base/Skeleton/User';

import {FETCH_COMMENTS} from '../../redux/comment.reducer';

class ListCard extends React.Component {
  componentWillMount() {
    if (!this.props.skeleton) {
      this.updateComments();
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.skeleton !== this.props.skeleton) {
      this.updateComments();
    }
  }

  updateComments() {
    this.props.fetchComments(this.props.list.id);

    if (this.props.list.list && this.props.list.list.length > 0) {
      this.props.list.list.forEach(el => {
        this.props.fetchComments(el._key + '-' + el.pid);
      });
    }
  }

  renderBookCover(id, img) {
    return img ? (
      <img alt="" src={'v1/image/' + img + '/250/200'} />
    ) : (
      <div
        className="list-card-coverTemplate"
        style={{background: toColor(id), height: '100%'}}
      >
        <div className="list-card-brick" />
        <div className="list-card-brick" />
        <div className="list-card-brick" />
      </div>
    );
  }

  render() {
    const {list, style, skeleton, className = ''} = this.props;
    let commentCount = 0;

    // if skelleton component is false - count comments
    if (!this.props.skeleton) {
      if (
        this.props.comments &&
        this.props.comments[list.id] &&
        this.props.comments[list.id].comments
      ) {
        // count comments for list
        commentCount = this.props.comments[list.id].comments.length;
        if (this.props.list.list && this.props.list.list.length > 0) {
          this.props.list.list.forEach(el => {
            if (this.props.comments[el._key + '-' + el.pid].comments) {
              // count comments for books in list
              commentCount += this.props.comments[el._key + '-' + el.pid]
                .comments.length;
            }
          });
        }
      }
    }

    // if no elements detected, show skeleton cards
    if (skeleton) {
      return (
        <div
          className={'ListCard__skeleton col-xs-12 pt1 mr1 ml1 ' + className}
        >
          <div className="ListCard__skeleton__cover" />
          <SkeletonText lines={3} color="#e9eaeb" />
          <SkeletonUser
            name={true}
            pulse={false}
            styles={{marginTop: '20px'}}
          />
        </div>
      );
    }

    return (
      <div className={'list-card col-xs-12 mr1 ml1 ' + className} style={style}>
        <Link href={`/lister/${list.id}`}>
          <div className="list-card-wrap">
            <div className="list-card-cover">
              {this.renderBookCover(list.id, list.image)}
            </div>
            <div className="list-card-summary">
              <Heading
                className="list-card-about mt2 mb1"
                Tag="h3"
                type="title"
              >
                <TruncateMarkup lines={3}>
                  <strong>
                    <div className="list-card-title">{list.title}</div>
                    <div className="list-card-description">
                      {list.description}
                    </div>
                  </strong>
                </TruncateMarkup>
              </Heading>
            </div>
            <div className="list-card-bottom">
              <ProfileImage
                key={'profile-img-' + list.id}
                user={this.props.profile}
                namePosition={'right'}
                type="list"
              />
              <div className="list-card-interactions">
                <span className="list-card-interaction">
                  <Icon className="md-18" name="chat_bubble_outline" /> ({
                    commentCount
                  })
                </span>
              </div>
            </div>
          </div>
        </Link>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  comments: state.comments
});

export const mapDispatchToProps = dispatch => ({
  fetchComments: id => dispatch({type: FETCH_COMMENTS, id})
});

export default connect(mapStateToProps, mapDispatchToProps)(ListCard);
