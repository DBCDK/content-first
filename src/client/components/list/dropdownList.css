.short-badge {
  letter-spacing: 2px;
}
.top-bar-dropdown-list--close-btn {
  position: absolute;
  cursor: pointer;
  width: 12px;
  top: 15x;
  left: 15px;
}
.top-bar-dropdown-list--content {
  position: fixed;
  background: var(--lys-graa);
  width: 350px;
  right: 10px;
  top: 80px;
  z-index: 1100;
  border: 1px solid rgba(0, 0, 0, 0.15);
  box-shadow: 0 2px 6px rgba(0, 0, 0, 0.175);
  font-family: var(--secondary-font);
  opacity: 1;
  visibility: visible;
  transition: all 0.15s ease-in-out;
  padding-top: 21px;
}
.slide-out {
  visibility: hidden;
  transition: all 0.15s ease-in-out;
  opacity: 0;
  right: -10px;
}
.top-bar-dropdown-list--empty-text {
  margin: 50px;
}
.top-bar-dropdown-list--elements {
  height: 450px;
  overflow-y: scroll;
  overflow-x: hidden;
  padding: 3px 0px;
}
.top-bar-dropdown-list-element {
  margin: 10px;
  margin: 10px 10px 0 10px;
  padding-left: 10px;
  padding-bottom: 10px;
}
.top-bar-dropdown-list-element:after {
  content: ' ';
  display: block;
  clear: both;
}
.top-bar-dropdown-list-element:last-child {
  border-bottom: 0;
}
.top-bar-dropdown-list-element--header a {
  font-family: var(--primary-font);
  color: var(--petrolium);
  font-weight: 600;
  font-size: 1.1em;
}
.top-bar-dropdown-list-element--cover-image {
  display: inline-block;
  float: left;
  padding: 2px;
}
.top-bar-dropdown-list-element--cover-image img {
  width: 40px;
}
.top-bar-dropdown-list-element--text {
  display: inline-block;
  float: left;
  width: 200px;
  margin-left: 1.5em;
  margin-top: -4px;
}
.top-bar-dropdown-list-element {
  position: relative;
  clear: both;
  overflow: hidden;
}
.top-bar-dropdown-list-element--close-btn {
  position: absolute;
  right: 5px;
  cursor: pointer;
  width: 11px;
}
.top-bar-dropdown-list-element--taxonomy-description {
  color: var(--petrolium);
  white-space: nowrap;
  overflow: hidden !important;
  text-overflow: ellipsis;
}
.top-bar-dropdown-list-element--origin {
  color: gray;
  font-size: 0.8em;
}
.top-bar-dropdown-list--more-elements {
  margin: 10px;
}
.top-bar-dropdown-list--more-elements-line {
  border-bottom: 1px solid lightgray;
  margin-top: -9px;
}
.top-bar-dropdown-list--more-elements-text {
  background: #fff;
  padding: 0 10px;
  font-weight: 600;
  font-size: 12px;
  color: gray;
}
.top-bar-dropdown-list--footer {
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  width: 80%;
  border-top: 1px solid lightgray;
  margin: auto;
  padding: 35px 0 23px 0;
  margin-top: 3px;
}
.top-bar-dropdown-list--more-btn {
  margin: auto;
  background: var(--concrete);
  color: var(--petrolium);
}
.top-bar-dropdown-list--merge-modal .work-list {
  margin-top: 20px;
}
.top-bar-dropdown-list--merge-modal .work-item-small {
  padding-bottom: 10px;
  padding-top: 10px;
  border-top: 1px solid var(--lightgrey);
}
.top-bar-dropdown-list--merge-modal .more-btn {
  margin-top: 30px;
}
.top-bar-dropdown-list-page .items {
  margin-top: 40px;
}
.top-bar-dropdown-list-page .items .item {
  border-bottom: 1px solid lightgray;
  padding: 10px;
  overflow: hidden;
}
.top-bar-dropdown-list-page .items .item:first-child {
  border-top: 1px solid lightgray;
}
.top-bar-dropdown-list-page .remove-btn {
  position: absolute;
  right: 0px;
  /*top: 10px;*/
  cursor: pointer;
  width: 10px;
}
.top-bar-dropdown-list-page .book-cover {
  height: 60px;
  float: left;
}
.top-bar-dropdown-list-page .item {
  font-family: var(--secondary-font);
}
.top-bar-dropdown-list-page .title {
  font-family: var(--primary-font);
  font-style: italic;
  font-weight: bold;
}
.top-bar-dropdown-list-page .creator {
  font-family: var(--primary-font);
  color: grey;
}
.top-bar-dropdown-list-page textarea {
  resize: none;
  width: 200px;
  margin-top: 10px;
}
.top-bar-dropdown-list-page .item form {
  width: 200px;
  margin-top: 10px;
}
.top-bar-dropdown-list-page textarea {
  resize: none;
}
.top-bar-dropdown-list-page .glyphicon {
  font-size: 11px;
}
.top-bar-dropdown-list-page .add-to-list > span,
.top-bar-dropdown-list-page .order-book > span {
  cursor: pointer;
}
.top-bar-dropdown-list-page .empty-list-text {
  font-family: var(--secondary-font);
  font-size: 24px;
  margin-top: 200px;
}

.dropdownlist-enter {
  opacity: 0;
  max-height: 0px;
}

.dropdownlist-enter.dropdownlist-enter-active {
  opacity: 1;
  transition: all 0.2s;
  max-height: 200px;
}
.dropdownlist-leave {
  opacity: 1;
  max-height: 200px;
}
.dropdownlist-leave.dropdownlist-leave-active,
.dropdownlist-leave.dropdownlist-leave-active * {
  padding-top: 0 !important;
  padding-bottom: 0 !important;
  margin-top: 0 !important;
  margin-bottom: 0 !important;
  border: 0;
  overflow: hidden;
}
.dropdownlist-leave.dropdownlist-leave-active {
  opacity: 0;
  max-height: 0px;
  transition: all 0.2s;
}
/* scrollbar in dropdownlist*/
.top-bar-dropdown-list--elements::-webkit-scrollbar {
  width: 8px;
}
.top-bar-dropdown-list--elements::-webkit-scrollbar-track {
  background: var(--concrete);
}
.top-bar-dropdown-list--elements::-webkit-scrollbar-thumb {
  background: var(--porcelain);
}
.top-bar-dropdown-list--elements::-webkit-scrollbar-thumb:hover {
  background: var(--alto);
}
.list-card-brick-small {
  background-color: #fff;
  width: 70%;
  height: 5px;
  border-radius: 1px;
  align-self: flex-start;
}
.list-card-coverTemplate-small {
  padding: 7px 8px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
}
.list-card-brick-small:nth-child(2) {
  align-self: flex-end;
}

/* Hacks for IE10 and above */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
  .top-bar-dropdown-list-element--cover-image img {
    max-width: 40px;
    width: auto;
  }
}
