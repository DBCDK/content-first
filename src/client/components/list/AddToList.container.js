import React from 'react';
import {connect} from 'react-redux';
import {addElementToList, storeList} from '../../redux/list.reducer';
import {OPEN_MODAL} from '../../redux/modal.reducer';
import BookSearchSuggester from './BookSearchSuggester';
import {ListItem} from './ListCreate.container';

export class AddToList extends React.Component {
  constructor() {
    super();
    this.state = {elementToAdd: null};
  }
  render() {
    const {list, allowAdd, addElement, requireLogin} = this.props;
    if (!allowAdd || !list) {
      return null;
    }

    const elementToAdd = this.state.elementToAdd;
    const alreadyInList = elementToAdd
      ? list.list.filter(item => item.book.pid === elementToAdd.book.pid)
          .length > 0
      : false;
    return (
      <div className="addbook">
        <h2 className="list-creator__headline mt2">Tilføj bøger til listen</h2>

        <BookSearchSuggester
          list={list}
          onSubmit={book => this.setState({elementToAdd: book})}
        />
        {elementToAdd &&
          !alreadyInList && (
            <div className="item mt4">
              <ListItem
                key={elementToAdd.book.pid}
                item={elementToAdd}
                onChange={(item, description) => {
                  item.description = description;
                  this.setState({elementToAdd: item});
                }}
              />
              <div className="text-right mt1">
                <span
                  className="btn btn-default"
                  onClick={() => {
                    this.setState({elementToAdd: null});
                  }}
                >
                  Fortryd
                </span>
                <span
                  className="btn btn-success ml1"
                  onClick={() => {
                    if (this.props.isLoggedIn) {
                      addElement(elementToAdd, list);
                      this.setState({elementToAdd: null});
                    } else {
                      requireLogin();
                    }
                  }}
                >
                  Tilføj
                </span>
              </div>
            </div>
          )}
        {alreadyInList && (
          <h4 className="mt2 mb2 text-center">
            <strong>{this.state.elementToAdd.book.title}</strong> kan ikke
            tilføjes, da den allerede er i listen
          </h4>
        )}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    allowAdd:
      ownProps.list.open ||
      ownProps.list.owner === state.userReducer.openplatformId,
    isLoggedIn: state.userReducer.isLoggedIn
  };
};
export const mapDispatchToProps = dispatch => ({
  addElement: async (book, list) => {
    await dispatch(addElementToList(book, list.id));
    dispatch(storeList(list.id));
  },
  requireLogin: () => {
    dispatch({
      type: OPEN_MODAL,
      modal: 'login',
      context: {
        title: 'LÆG I LISTE',
        reason: 'Du skal logge ind for at lægge bøger i en liste.'
      }
    });
  }
});
export default connect(mapStateToProps, mapDispatchToProps)(AddToList);
