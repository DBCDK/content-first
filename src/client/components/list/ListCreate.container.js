import React from 'react';
import {connect} from 'react-redux';
import {
  LIST_LOAD_REQUEST,
  addElementToList,
  removeElementFromList,
  updateList,
  storeList,
  getListById,
  addList,
  ADD_LIST_IMAGE
} from '../../redux/list.reducer';
import {createListLocation} from '../../utils/requestLists';
import DragableList from './ListDrag.component';
import Textarea from 'react-textarea-autosize';
import {HISTORY_REPLACE} from '../../redux/middleware';
import BookSearchSuggester from './BookSearchSuggester';
import BookCover from '../general/BookCover.component';
import Link from '../general/Link.component';
import ImageUpload from '../general/ImageUpload.component';
const ListDetails = ({
  id,
  title,
  description,
  template,
  hasError,
  onChange,
  addImage,
  imageError,
  imageIsLoading,
  image
}) => (
  <div className="list-details">
    <div className="form-group">
      <span className={`required ${!title && hasError ? 'has-error' : ''}`}>
        <input
          className="form-control"
          type="text"
          name="list-title"
          placeholder="Giv din liste en titel"
          onChange={e => onChange({title: e.currentTarget.value})}
          value={title}
        />
        {!title && hasError ? (
          <div className="alert alert-danger">Din liste skal have en titel</div>
        ) : (
          ''
        )}
      </span>
      <Textarea
        className="form-control list-details__description"
        name="list-description"
        placeholder="Skriv lidt om din liste"
        onChange={e => onChange({description: e.currentTarget.value})}
        value={description}
      />
      <div>
        <span className="ml1">Skal vises som</span>
        <select
          value={template || 'simple'}
          onChange={e => onChange({template: e.currentTarget.value})}
          className="form-control ml1"
          style={{width: 'auto', display: 'inline-block'}}
        >
          <option value="simple">simpel liste</option>
          <option value="circle">visuel liste</option>
        </select>
      </div>

      <div className="mt1 text-left">
        <ImageUpload
          className="mt1"
          icon="glyphicon-picture"
          error={imageError}
          style={{borderRadius: '5%'}}
          loading={imageIsLoading}
          previewImage={image ? `/v1/image/${image}/150/150` : null}
          onFile={img => {
            addImage(id, img);
          }}
        />
      </div>
    </div>
  </div>
);

const ListItem = ({item, onChange}) => (
  <div key={item.book.pid} className="flex list-item">
    <BookCover className="list-item__img" book={item.book} />
    <div className="list-item__details">
      <div className="list-item__title">{item.book.title}</div>
      <div className="list-item__author">{item.book.creator}</div>
      <Textarea
        onChange={e => onChange(item, e.currentTarget.value)}
        value={item.description || ''}
        className="form-control"
        name="list-description"
        placeholder="Skriv en kommentar til bogen"
      />
    </div>
  </div>
);

const ListBooks = props => (
  <div className="list-drag">
    <BookSearchSuggester
      list={props.list}
      onSubmit={book => props.addElementToList(book, props.list.data.id)}
    />
    <DragableList
      list={props.list.data.list}
      renderListItem={ListItem}
      onUpdate={updatedList => {
        props.updateList({...props.list.data, list: updatedList});
      }}
      onRemove={book => props.removeElementFromList(book, props.list.data.id)}
    />
  </div>
);

export class ListCreator extends React.Component {
  constructor() {
    super();
    this.state = {
      hasError: false
    };
  }
  async componentWillMount() {
    // check if we need to create a new list
    if (!this.props.id && this.props.createList) {
      this.props.createList();
    }
  }
  componentWillUnmount() {
    // reset any unsaved changes
    // for now we just reload users lists from backend
    // a client side undo mechanism would be more efficient
    this.props.loadLists();
  }
  async onSubmit(e) {
    e.preventDefault();
    if (!this.props.currentList.data.title) {
      this.setState({hasError: true});
      window.scrollTo(0, 0);
      return;
    }
    await this.props.storeList(this.props.currentList);
  }
  onChange(currentList) {
    this.props.updateList({...this.props.currentList.data, ...currentList});
  }
  setStatus() {
    const currentList = this.props.currentList;
    this.props.updateList({
      id: currentList.data.id,
      public: !currentList.data.public
    });
  }
  render() {
    if (!this.props.currentList) {
      return null;
    }
    const isNew = this.props.currentList.data.created_epoch ? false : true;
    return (
      <div className="list-creator">
        <h1 className="list-creator__headline">
          {isNew ? 'Opret liste' : 'Redigér liste'}
        </h1>
        <div className="row">
          <div className="col-xs-8">
            <form className="mb4" onSubmit={e => this.onSubmit(e)}>
              <ListDetails
                id={this.props.currentList.data.id}
                hasError={this.state.hasError}
                title={this.props.currentList.data.title}
                description={this.props.currentList.data.description}
                onChange={e => this.onChange(e)}
                addImage={this.props.addImage}
                image={this.props.currentList.data.image}
                imageError={this.props.currentList.data.imageError}
                imageIsLoading={this.props.currentList.data.imageIsLoading}
                template={this.props.currentList.data.template}
              />
              <h2 className="list-creator__headline">
                Tilføj bøger til listen
              </h2>
              <ListBooks
                updateList={this.props.updateList}
                addElementToList={this.props.addElementToList}
                removeElementFromList={this.props.removeElementFromList}
                list={this.props.currentList}
              />
              <div className="list-creator__publication">
                <label htmlFor="public">
                  <input
                    id="public"
                    name="public"
                    type="checkbox"
                    checked={this.props.currentList.data.public || false}
                    onClick={() => this.setStatus()}
                  />
                  <span /> Skal listen være offentlig?
                </label>
              </div>
              <div className="list-creator__submit text-right">
                <button className="btn btn-primary" type="submit">
                  Gem liste
                </button>
              </div>
              <Link href="/profile" replace={true}>
                {isNew
                  ? 'Fortryd oprettelse af liste'
                  : 'Fortryd redigering af liste'}
              </Link>
            </form>
          </div>
          <div className="col-xs-4" />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    currentList: getListById(state.listReducer, ownProps.id)
  };
};
export const mapDispatchToProps = dispatch => ({
  addImage: (id, image) => dispatch({type: ADD_LIST_IMAGE, image, id}),
  updateList: data => dispatch(updateList(data)),
  storeList: async list => {
    await dispatch(storeList(list.data.id));
    dispatch({type: HISTORY_REPLACE, path: '/lister'});
  },
  addElementToList: (book, id) => dispatch(addElementToList(book, id)),
  removeElementFromList: (book, id) =>
    dispatch(removeElementFromList(book, id)),
  loadLists: () => dispatch({type: LIST_LOAD_REQUEST}),
  createList: async () => {
    const {id} = await createListLocation();
    dispatch(addList({id}));
    dispatch({
      type: HISTORY_REPLACE,
      path: `/lister/${id}/rediger`
    });
  }
});
export default connect(mapStateToProps, mapDispatchToProps)(ListCreator);
