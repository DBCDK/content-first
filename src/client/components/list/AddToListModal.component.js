import React from 'react';
import Modal from '../general/Modal.component';
import WorkItemSmall from '../work/WorkItemSmall.component';
import {SYSTEM_LIST} from '../../redux/list.reducer';

const defaultState = {
  comment: '',
  list: '',
  listName: ''
};

class AddToListModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = Object.assign({}, defaultState);
  }

  render() {
    const customLists = this.props.lists.filter(l => l.type !== SYSTEM_LIST);

    return (
      <Modal
        className="add-to-list--modal"
        show={this.props.show}
        header={'GEM I LISTE'}
        onClose={() => {
          this.props.onClose();
          this.setState(Object.assign({}, defaultState));
        }}
        onDone={() => {
          this.props.onDone(this.props.work, this.state.comment, this.state.list);
          this.setState(Object.assign({}, defaultState));
        }}
        doneText="JA TAK, GEM NU">
        <div className="row">
          <strong className="col-xs-12">Hvilken liste vil du gemme i?</strong>
        </div>
        <div className="row">
          <div className="col-xs-6">
            <div className="list-overview">
              {customLists.map(l => {
                return <div key={l.id}>
                  <input
                    type="radio"
                    name="list"
                    checked={this.state.list.id === l.id}
                    onChange={() => this.setState({list: l})} />{l.title}
                </div>;
              })}
            </div>
            <div className="add-list">
              <form
                onSubmit={e => {
                  if (this.state.listName) {
                    this.props.onAddList(this.state.listName);
                    this.setState({listName: ''});
                  }
                  e.preventDefault();
                }}>
                <input
                  type="text"
                  name="add-list"
                  placeholder="Opret ny liste"
                  value={this.state.listName}
                  onChange={e => this.setState({listName: e.target.value})}/>
                <input className="add-list--btn text-center" type="submit" value="+" />
              </form>
            </div>
          </div>
          <div className="col-xs-6">
            {this.props.work && <WorkItemSmall work={this.props.work}/>}
            <textarea
              className="comment"
              placeholder="Skriv evt. en kommentar til bogen"
              value={this.state.comment}
              onChange={(e) => this.setState({comment: e.target.value})}>
            </textarea>
          </div>
        </div>

      </Modal>
    );
  }
}
export default AddToListModal;
