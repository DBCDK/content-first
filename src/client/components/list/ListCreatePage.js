import React from 'react';
import {connect} from 'react-redux';
import {addList, CUSTOM_LIST} from '../../redux/list.reducer';
import {HISTORY_REPLACE} from '../../redux/middleware';
import Spinner from '../general/Spinner.component';

/*
* Will create a new list object and redirect to listpage
*/
export class ListCreator extends React.Component {
  componentDidMount() {
    this.props.createList();
  }

  render() {
    return (
      <div className="d-flex justify-content-center">
        <Spinner size="30px" className="mt-5" />
      </div>
    );
  }
}

const mapStateToProps = () => ({});
export const mapDispatchToProps = dispatch => ({
  createList: async () => {
    await dispatch(
      addList({type: CUSTOM_LIST, isNew: true}, list => {
        dispatch({
          type: HISTORY_REPLACE,
          path: `/lister/${list._id}`
        });
      })
    );
  }
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListCreator);
