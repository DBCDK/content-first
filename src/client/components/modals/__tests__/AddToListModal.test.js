import React from 'react';
import {Provider} from 'react-redux';
import AddToListModal from '../AddToListModal.container';
import renderer from 'react-test-renderer';
import createStore from '../../../redux/Store';
import {ADD_LIST} from '../../../redux/list.reducer';
import {createTestList, createTestElement} from '../../../utils/testHelper';

// for ref in AddToListModal not to be null
function createNodeMock() {
  return {};
}

describe('AddToListModal', () => {
  const store = createStore();
  store.dispatch({type: ADD_LIST, list: createTestList(1)});
  store.dispatch({type: ADD_LIST, list: createTestList(2)});
  const tree = renderer.create(
    <Provider store={store}>
      <AddToListModal work={createTestElement(1)} />
    </Provider>,
    {createNodeMock}
  );
  test('renders lists from store, first is selected', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });
  test('newly added list is selected', () => {
    store.dispatch({type: ADD_LIST, list: createTestList(3)});
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
