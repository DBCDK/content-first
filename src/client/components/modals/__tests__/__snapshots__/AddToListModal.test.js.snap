// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToListModal newly added list is selected 1`] = `
<div
  className="modal-backdrop"
>
  <div
    className="modal-window text-left add-to-list--modal"
  >
    <img
      alt="luk"
      className="modal-window--close-btn"
      onClick={[Function]}
      src="Kryds.svg"
    />
    <div
      className="modal-window--header text-center"
    >
      <h3>
        GEM I LISTE
      </h3>
    </div>
    <div
      className="modal-window--content"
    >
      <div
        className="row"
      >
        <strong
          className="col-xs-12"
        >
          Hvilken liste vil du gemme i?
        </strong>
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-6"
        >
          <div
            className="list-overview"
          >
            <div>
              <input
                checked={false}
                name="list"
                onChange={[Function]}
                type="radio"
              />
              title1
            </div>
            <div>
              <input
                checked={false}
                name="list"
                onChange={[Function]}
                type="radio"
              />
              title2
            </div>
            <div>
              <input
                checked={true}
                name="list"
                onChange={[Function]}
                type="radio"
              />
              title3
            </div>
          </div>
          <div
            className="add-list"
          >
            <form
              onSubmit={[Function]}
            >
              <input
                name="add-list"
                onChange={[Function]}
                placeholder="Opret ny liste"
                type="text"
                value=""
              />
              <input
                className="add-list--btn text-center"
                type="submit"
                value="+"
              />
            </form>
          </div>
        </div>
        <div
          className="col-xs-6"
        >
          <div
            className="work-item-small clearfix"
          >
            <div
              className="col-xs-3"
            >
              <div
                className="cover-image-wrapper"
              >
                <img
                  alt=""
                  onError={[Function]}
                  src="https://metakompasset.demo.dbc.dk/api/cover/pid1"
                />
              </div>
            </div>
            <div
              className="col-xs-9 info"
            >
              <div
                className="title"
              >
                some title1
              </div>
              <div
                className="creator"
              >
                some creator1
              </div>
            </div>
          </div>
          <textarea
            className="comment"
            onChange={[Function]}
            placeholder="Skriv evt. en kommentar til bogen"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="modal-window--buttons text-center"
    >
      <span
        className="btn btn-success"
        onClick={[Function]}
      >
        JA TAK, GEM NU
      </span>
    </div>
  </div>
</div>
`;

exports[`AddToListModal renders lists from store, first is selected 1`] = `
<div
  className="modal-backdrop"
>
  <div
    className="modal-window text-left add-to-list--modal"
  >
    <img
      alt="luk"
      className="modal-window--close-btn"
      onClick={[Function]}
      src="Kryds.svg"
    />
    <div
      className="modal-window--header text-center"
    >
      <h3>
        GEM I LISTE
      </h3>
    </div>
    <div
      className="modal-window--content"
    >
      <div
        className="row"
      >
        <strong
          className="col-xs-12"
        >
          Hvilken liste vil du gemme i?
        </strong>
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-6"
        >
          <div
            className="list-overview"
          >
            <div>
              <input
                checked={true}
                name="list"
                onChange={[Function]}
                type="radio"
              />
              title1
            </div>
            <div>
              <input
                checked={false}
                name="list"
                onChange={[Function]}
                type="radio"
              />
              title2
            </div>
          </div>
          <div
            className="add-list"
          >
            <form
              onSubmit={[Function]}
            >
              <input
                name="add-list"
                onChange={[Function]}
                placeholder="Opret ny liste"
                type="text"
                value=""
              />
              <input
                className="add-list--btn text-center"
                type="submit"
                value="+"
              />
            </form>
          </div>
        </div>
        <div
          className="col-xs-6"
        >
          <div
            className="work-item-small clearfix"
          >
            <div
              className="col-xs-3"
            >
              <div
                className="cover-image-wrapper"
              >
                <img
                  alt=""
                  onError={[Function]}
                  src="https://metakompasset.demo.dbc.dk/api/cover/pid1"
                />
              </div>
            </div>
            <div
              className="col-xs-9 info"
            >
              <div
                className="title"
              >
                some title1
              </div>
              <div
                className="creator"
              >
                some creator1
              </div>
            </div>
          </div>
          <textarea
            className="comment"
            onChange={[Function]}
            placeholder="Skriv evt. en kommentar til bogen"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="modal-window--buttons text-center"
    >
      <span
        className="btn btn-success"
        onClick={[Function]}
      >
        JA TAK, GEM NU
      </span>
    </div>
  </div>
</div>
`;
